tests/core/celery/tasks/test_generate_inventory_and_hold_reports.py

```diff
@@ -0,0 +1,288 @@
+        assert row["language"] == language
+        assert row["publisher"] == publisher
+        assert row["audience"] == "young adult"
+        assert row["genres"] == "genre_a,genre_z"
+        assert row["format"] == edition.BOOK_MEDIUM
+        assert row["data_source"] == data_source
+        assert row["collection_name"] == collection_name
+        assert float(row["days_remaining_on_license"]) == float(days_remaining)
+        assert row["shared_active_loan_count"] == "0"
+        assert row["library_active_loan_count"] == "0"
+        assert row["remaining_loans"] == str(checkouts_left)
+        assert row["allowed_concurrent_users"] == str(terms_concurrency)
+        assert expiration.strftime("%Y-%m-%d %H:%M:%S.%f") in row["license_expiration"]
+
+    holds_report_key = [x for x in attachments.keys() if "holds" in x][0]
+    assert holds_report_key
+    assert "test_library" in holds_report_key
+    holds_report_value = attachments[holds_report_key]
+    assert holds_report_value
+    holds_report_csv = list(csv.DictReader(open(holds_report_value)))
+    assert len(holds_report_csv) == 1
+
+    for row in holds_report_csv:
+        assert row["title"] == title
+        assert row["author"] == author
+        assert row["identifier"]
+        assert row["language"] == language
+        assert row["publisher"] == publisher
+        assert row["audience"] == "young adult"
+        assert row["genres"] == "genre_a,genre_z"
+        assert row["format"] == edition.BOOK_MEDIUM
+        assert row["data_source"] == data_source
+        assert row["collection_name"] == collection_name
+        assert int(row["shared_active_hold_count"]) == shared_patrons_in_hold_queue
+        assert int(row["library_active_hold_count"]) == 3
+
+    # clean up files
+    for f in attachments.values():
+        os.remove(f)
+
+
+def create_test_opds_collection(
+    collection_name: str,
+    data_source: str,
+    db: DatabaseTransactionFixture,
+    library: Library,
+):
+    settings = OPDSImporterSettings(
+        include_in_inventory_report=True,
+        external_account_id="http://opds.com",
+        data_source=data_source,
+    )
+    collection = db.collection(name=collection_name, settings=settings.dict())
+    collection.libraries = [library]
+    return collection
+
+
+def test_generate_inventory_and_hold_reports_task(
+    db: DatabaseTransactionFixture,
+    services_fixture: ServicesFixture,
+    celery_fixture: CeleryFixture,
+):
+    library = db.library(short_name="test_library")
+    # there must be at least one opds collection associated with the library for this to work
+    create_test_opds_collection("c1", "d1", db, library)
+    send_email_mock = create_autospec(
+        services_fixture.services.email.container.send_email
+    )
+    services_fixture.services.email.container.send_email = send_email_mock
+    generate_inventory_and_hold_reports.delay(library.id, "test@email").wait()
+    send_email_mock.assert_called_once()
```

Comment:
You could rewrite the mocks here, so that they check on the mock email sender provided by the services fixture. That way we know that the correct `send_email` gets passed in from the container. That way we don't run into a situation like we have here where the mock provides something the real container doesn't have.

```python
def test_generate_inventory_and_hold_reports_task(
    db: DatabaseTransactionFixture,
    services_fixture: ServicesFixture,
    celery_fixture: CeleryFixture,
):
    library = db.library(short_name="test_library")
    # there must be at least one opds collection associated with the library for this to work
    create_test_opds_collection("c1", "d1", db, library)
    generate_inventory_and_hold_reports.delay(library.id, "test@email").wait()
    services_fixture.email_fixture.mock_emailer.send.assert_called_once()
    assert "Inventory and Holds Reports" in services_fixture.email_fixture.mock_emailer.send.call_args.kwargs["subject"]
    assert services_fixture.email_fixture.mock_emailer.send.call_args.kwargs["receivers"] == ["test@email"]
```

Reply:
Nice to add some assertions as well about what gets passed into `send` to make sure the email address etc is making the full trip into the emailer.